pyts/classification/bosssp.py

```diff
@@ -0,0 +1,270 @@
+        - 'quantile': All bins in each sample have the same number of points
+        - 'normal': Bin edges are quantiles from a standard normal distribution
+        - 'entropy': Bin edges are computed using information gain
+
+    alphabet : None, 'ordinal' or array-like, shape = (n_bins,)
+        Alphabet to use. If None, the first `n_bins` letters of the Latin
+        alphabet are used.
+
+    numerosity_reduction : bool (default = True)
+        If True, delete sample-wise all but one occurence of back to back
+        identical occurences of the same words.
+
+    use_idf : bool (default = True)
+        Enable inverse-document-frequency reweighting.
+
+    smooth_idf : bool (default = False)
+        Smooth idf weights by adding one to document frequencies, as if an
+        extra document was seen containing every term in the collection
+        exactly once. Prevents zero divisions.
+
+    sublinear_tf : bool (default = True)
+        Apply sublinear tf scaling, i.e. replace tf with 1 + log(tf).
+
+    level : integer (default = 3)
+         Number of times the series is being divided. Maximum of 3.
+
+    References
+    ----------
+    .. [1] James Large et al., “On time series classification with
+           dictionary-based classifiers”. Intelligent Data Analysis 23.5 (2019)
+
+    """
+
+    def __init__(self, word_size=4, n_bins=4, window_size=10, window_step=1,
+                 anova=False, drop_sum=False, norm_mean=False, norm_std=False,
+                 strategy='quantile', alphabet=None,
+                 numerosity_reduction=True, use_idf=True, smooth_idf=False,
+                 sublinear_tf=True, level=3):
+        self.word_size = word_size
+        self.n_bins = n_bins
+        self.window_size = window_size
+        self.window_step = window_step
+        self.anova = anova
+        self.drop_sum = drop_sum
+        self.norm_mean = norm_mean
+        self.norm_std = norm_std
+        self.strategy = strategy
+        self.alphabet = alphabet
+        self.numerosity_reduction = numerosity_reduction
+        self.use_idf = use_idf
+        self.smooth_idf = smooth_idf
+        self.sublinear_tf = sublinear_tf
+        self.level = level
+
+    def fit(self, X, y):
+        """Compute the document-term matrix.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_timestamps)
+            Training vector.
+
+        y : array-like, shape = (n_samples,)
+            Class labels for each data sample.
+
+        Returns
+        -------
+        self : object
+
+        """
+        length = len(X)
+        words = self._boss_word_extractor(X, y)
+        if self.level >= 2:
+            words_2 = self._boss_word_extractor(X[:length//2],
+                                                y[:length//2])
+            words_3 = self._boss_word_extractor(X[length//2:],
+                                                y[:length//2:])
+            words = words + words_2 + words_3
```

Comment:
What are concatenated in BOSSSP are the histograms (and not the list of words). For instance, if `level=3`, the final histogram is the concatenation of 7 histograms. You need to first compute the histogram for each `words_i` list, then concatenate the histogram, instead of concatenating first the `words_i` lists and then computing the histogram.